File: server/create_embeddings_table.ts
import { db, pool } from './db';
import { sql } from 'drizzle-orm';

/**
 * Script to create embeddings table with vector support
 * This is needed because we're using the pgvector extension which requires special SQL
 */
async function createEmbeddingsTable() {
  try {
    console.log('Creating embeddings table with vector support...');
    
    // First ensure the vector extension is installed
    await db.execute(sql`CREATE EXTENSION IF NOT EXISTS vector;`);
    
    // Create the embeddings table with a vector column for storing embeddings
    const createTableSQL = sql`
      CREATE TABLE IF NOT EXISTS embeddings (
        id SERIAL PRIMARY KEY,
        category TEXT NOT NULL,
        requirement TEXT NOT NULL,
        response TEXT NOT NULL,
        reference TEXT,
        payload TEXT NOT NULL,
        embedding VECTOR(1536) NOT NULL,
        timestamp TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP NOT NULL
      );
    `;
    
    await db.execute(createTableSQL);
    
    // Create an index for faster vector similarity searches
    await db.execute(sql`
      CREATE INDEX IF NOT EXISTS embeddings_vector_idx 
      ON embeddings 
      USING ivfflat (embedding vector_cosine_ops)
      WITH (lists = 100);
    `);
    
    console.log('Successfully created embeddings table with vector support!');
  } catch (error) {
    console.error('Error creating embeddings table:', error);
    throw error;
  } finally {
    // Don't close the pool here if it's shared with other parts of the application
    // await pool.end();
  }
}

// Export the function so it can be called from other files
export { createEmbeddingsTable };

// Run the migration if this script is called directly
if (require.main === module) {
  createEmbeddingsTable()
    .then(() => {
      console.log('Migration completed successfully!');
      process.exit(0);
    })
    .catch((error) => {
      console.error('Migration failed:', error);
      process.exit(1);
    });
}

File: server/create_embeddings_table.py
#!/usr/bin/env python3
"""
Script to create embeddings table with vector support
This is needed because we're using the pgvector extension which requires special SQL
"""

import os
import sys
import psycopg2

def get_database_connection():
    """Get a connection to the PostgreSQL database."""
    try:
        # Get connection string from environment variable
        db_url = os.environ.get('DATABASE_URL')
        if not db_url:
            print("DATABASE_URL environment variable not set")
            sys.exit(1)
        
        # Connect to the database
        conn = psycopg2.connect(db_url)
        print("Successfully connected to PostgreSQL database")
        return conn
    except Exception as e:
        print(f"Error connecting to database: {str(e)}")
        sys.exit(1)

def create_embeddings_table():
    """Create the embeddings table with vector support."""
    conn = get_database_connection()
    cursor = conn.cursor()
    
    try:
        # Check if pgvector extension exists
        cursor.execute("SELECT EXISTS (SELECT 1 FROM pg_extension WHERE extname = 'vector');")
        extension_exists = cursor.fetchone()[0]
        
        if not extension_exists:
            print("Creating pgvector extension...")
            cursor.execute("CREATE EXTENSION IF NOT EXISTS vector;")
            print("pgvector extension created successfully")
        else:
            print("pgvector extension already exists")
        
        # Create embeddings table with vector column
        print("Creating embeddings table...")
        cursor.execute("""
        CREATE TABLE IF NOT EXISTS embeddings (
            id SERIAL PRIMARY KEY,
            category TEXT NOT NULL,
            requirement TEXT NOT NULL,
            response TEXT,
            reference TEXT,
            payload JSONB,
            embedding vector(1536),
            created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
        );
        """)
        print("Embeddings table created successfully")
        
        # Create index for faster vector similarity search
        print("Creating vector index for similarity search...")
        cursor.execute("""
        CREATE INDEX IF NOT EXISTS embeddings_embedding_idx 
        ON embeddings 
        USING ivfflat (embedding vector_cosine_ops)
        WITH (lists = 100);
        """)
        print("Vector index created successfully")
        
        # Create text search indices for traditional search
        print("Creating text search indices...")
        cursor.execute("CREATE INDEX IF NOT EXISTS embeddings_category_idx ON embeddings (category);")
        cursor.execute("CREATE INDEX IF NOT EXISTS embeddings_requirement_idx ON embeddings USING GIN (to_tsvector('english', requirement));")
        print("Text search indices created successfully")
        
        # Commit the changes
        conn.commit()
        print("All changes committed successfully")
        
    except Exception as e:
        conn.rollback()
        print(f"Error creating embeddings table: {str(e)}")
        sys.exit(1)
    finally:
        cursor.close()
        conn.close()

if __name__ == "__main__":
    print("Starting to create embeddings table...")
    create_embeddings_table()
    print("Script completed successfully")